infobind/cache.go

```diff
@@ -0,0 +1,256 @@
+
+	fs := rs.Fields()
+	chk := rs.NewChunk()
+	for {
+		err = rs.Next(context.TODO(), chk)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if chk.NumRows() == 0 {
+			return nil
+		}
+		it := chunk.NewIterator4Chunk(chk)
+		for row := it.Begin(); row != it.End(); row = it.Next() {
+			record, err := decodeBindTableRow(row, fs)
+			if err != nil {
+				log.Errorf("row decode error %s", err)
+				continue
+			}
+			log.Infof("record %v", record)
+			err = bc.appendNode(h.Ctx, record, h.Parser)
+			if err != nil {
+				continue
+			}
+
+			if record.UpdateTime.Compare(h.LastUpdateTime) == 1 {
+				h.LastUpdateTime = record.UpdateTime
+			}
+		}
+		chk = chunk.Renew(chk, h.Ctx.GetSessionVars().MaxChunkSize)
+	}
+}
+
+//Update update the HandleUpdater's bindCache if tidb first startup,the fullLoad is true,otherwise fullLoad is false
+func (h *HandleUpdater) Update(fullLoad bool) error {
+	var (
+		err error
+		sql string
+	)
+	bc := h.Get()
+	if fullLoad {
+		sql = fmt.Sprintf("select * from mysql.bind_info")
+	} else {
+		sql = fmt.Sprintf("select * from mysql.bind_info where update_time > \"%s\"", h.LastUpdateTime.String())
+	}
+	err = h.loadDiff(sql, bc)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	h.bind.Store(bc)
+	bc.Display()
+	return nil
+}
+
+func parseSQL(sctx sessionctx.Context, parser *parser.Parser, sql string, charset string, collation string) ([]ast.StmtNode, []error, error) {
+	return parser.Parse(sql, charset, collation)
+}
+
+func decodeBindTableRow(row chunk.Row, fs []*ast.ResultField) (bindRecord, error) {
+	var value bindRecord
+	for i, f := range fs {
+		switch {
+		case f.ColumnAsName.L == "original_sql":
+			value.OriginalSQL = row.GetString(i)
+		case f.ColumnAsName.L == "bind_sql":
+			value.BindSQL = row.GetString(i)
+		case f.ColumnAsName.L == "default_db":
+			value.Db = row.GetString(i)
+		case f.ColumnAsName.L == "status":
+			value.Status = row.GetInt64(i)
+		case f.ColumnAsName.L == "create_time":
+			var err error
+			value.CreateTime = row.GetTime(i)
+			if err != nil {
+				return value, errors.Trace(err)
+			}
+		case f.ColumnAsName.L == "update_time":
+			var err error
+			value.UpdateTime = row.GetTime(i)
+			if err != nil {
+				return value, errors.Trace(err)
+			}
+		case f.ColumnAsName.L == "charset":
+			var err error
+			value.Charset = row.GetString(i)
+			if err != nil {
+				return value, errors.Trace(err)
+			}
+		case f.ColumnAsName.L == "collation":
+			var err error
+			value.Collation = row.GetString(i)
+			if err != nil {
+				return value, errors.Trace(err)
+			}
+		}
+	}
+	return value, nil
+}
+
+func (b *BindCache) appendNode(sctx sessionctx.Context, value bindRecord, sparser *parser.Parser) error {
+	hash := parser.Digest(value.OriginalSQL)
+	if value.Status == 0 {
+		if bindArr, ok := b.Cache[hash]; ok {
+			if len(bindArr) == 1 {
+				if bindArr[0].Db == value.Db {
+					delete(b.Cache, hash)
+				}
+				return nil
+			}
+			for idx, v := range bindArr {
```

Comment:
You should loop it from the end? Or after it is removed, the index won't be correct for latter ones.

Reply:
I have done the test case like this:
`func main(){
	var arr []int

	for i:=0 ; i<10 ; i++ {
		arr = append(arr, i)
	}

	for idx,value := range arr {
		if value == 5 {
			arr = append(arr[:idx], arr[idx+1:]...)
		}
	}

	fmt.Println(arr)
}`

the output :
`[0 1 2 3 4 6 7 8 9]`

So I think this code is right.

Reply:
You can try if there are several matched items, not just one.

Reply:
The originSql and Db both match only occur once, because 
1. the `parser.Digest` use MD5, so we can think the hash match, the originSql is matched nearly.
2. the Db use string compare

Reply:
I think it cannot be guaranteed, and if it happens, it is hard to diagnose the problem.

Reply:
ok